Tests/test_file.py

```diff
@@ -13,827 +13,831 @@
+                        ("\r\n\r\n", "\r\r\n\r\r\n"),
+                        ("\n\r\n\r", "\r\n\r\r\n\r"),
+                        ("The quick brown fox", "The quick brown fox"),
+                        ("The \rquick\n brown fox\r\n", "The \rquick\r\n brown fox\r\r\n"),
+                        ("The \r\rquick\r\n\r\n brown fox", "The \r\rquick\r\r\n\r\r\n brown fox"))
+
+        # Test a specific read mode pattern.
+        def test_read_pattern(pattern):
+            # Write the initial data to disk using binary mode (we test this
+            # functionality earlier so we're satisfied it gets there unaltered).
+            with file(self.temp_file, "wb") as f:
+                f.write(pattern[0])
+
+            # Read the data back in each read mode, checking that we get the correct
+            # transform each time.
+            for mode in range(3):
+                test_read_mode(pattern, mode)
+
+        # Test a specific read mode pattern for a given reading mode.
+        def test_read_mode(pattern, mode):
+            # Read the data back from disk using the given read mode.
+            with file(self.temp_file, self.read_modes[mode][1]) as f:
+                contents = f.read()
+
+            # Check it equals what we expected for this mode.
+            self.assertTrue(contents == pattern[mode])
+
+        # Test a specific write mode pattern.
+        def test_write_pattern(pattern):
+            for mode in range(2):
+                test_write_mode(pattern, mode)
+
+        # Test a specific write mode pattern for a given write mode.
+        def test_write_mode(pattern, mode):
+            # Write the raw data using the given mode.
+            with file(self.temp_file, self.write_modes[mode][1]) as f:
+                f.write(pattern[0])
+
+            # Read the data back in using binary mode (we tested this gets us back
+            # unaltered data earlier).
+            with file(self.temp_file, "rb") as f:
+                contents = f.read()
+
+            # Check it equals what we expected for this mode.
+            self.assertTrue(contents == pattern[mode])
+
+        # Run through the read and write mode tests for all patterns.
+        def test_patterns():
+            for pattern in read_patterns:
+                test_read_pattern(pattern)
+            for pattern in write_patterns:
+                test_write_pattern(pattern)
+
+        # Actually run the pattern mode tests.
+        test_patterns()
+
+    # Now some tests of read(size).
+    # Test data is in the following format: ("raw data", read_size, (binary mode result strings) (binary mode result tell() result)
+    #                                                               (text mode result strings) (text mode result tell() result)
+    #                                                               (universal mode result strings) (universal mode result tell() results)
+
+    @unittest.skipUnless(is_cli, 'IronPython specific test')
```

Comment:
No, I just maintained what is there on this one. I don't see us running most of these tests under CPython, but I guess its possible.